submission/distributor.go

```diff
@@ -68,12 +74,68 @@ func (d *Distributor) Run(ctx context.Context) {
 	}
 }
 
-// refreshRoots requests roots from Logs and updates local copy if necessary.
+// refreshRoots requests roots from Logs and updates local copy.
 func (d *Distributor) refreshRoots(ctx context.Context) {
-	// TODO(Mercurrent) add implementation.
+	type RootsResult struct {
+		LogURL string
+		Roots  []*x509.Certificate
+		Err    error
+	}
+	ch := make(chan RootsResult, len(d.logClients))
+
+	rctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	for logURL, lc := range d.logClients {
+		go func(logURL string, lc client.AddLogClient) {
+			res := RootsResult{LogURL: logURL}
+
+			roots, err := lc.GetAcceptedRoots(rctx)
+			if err != nil {
+				res.Err = fmt.Errorf("%s: couldn't collect roots. %s", logURL, err)
+				ch <- res
+				return
+			}
+			for _, r := range roots {
+				parsed, err := x509.ParseCertificate(r.Data)
+				if err != nil {
+					res.Err = fmt.Errorf("%s: unable to parse root cert: %s", logURL, err)
+					continue
+				}
+				res.Roots = append(res.Roots, parsed)
+			}
+			ch <- res
+		}(logURL, lc)
+	}
+
+	// Collect get-roots results for every Log-client.
+	freshRoots := make(map[string][]*x509.Certificate)
+	for range d.logClients {
+		r := <-ch
+		// update roots for successful Log-requests only.
+		if r.Err != nil {
+			log.Println(r.Err)
+		} else {
+			freshRoots[r.LogURL] = r.Roots
+		}
+	}
+
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	for logURL, r := range freshRoots {
+		d.logRoots[logURL] = r
```

Comment:
I assume you're doing this, rather than just `d.logRoots = r`, so that anyone holding a copy of the `d.logRoots` pointer doesn't get left referencing an old copy of the map? If that's not a concern, then you could defer locking `d.mu` until after this.

Reply:
I'm trying to make roots-update atomic for r.logRoots.

Reply:
Ah the locking is meant to encompass both changing `d.logRoots` and `d.rootPool` so that they both update at the same time. Ok, SGTM - leave the locking where it is. I think you can still replace this loop with `d.logRoots = freshRoots` though, purely for conciseness.